### PR TITLE
Frontend setup

### PR DESCRIPTION
**Pull request by rperryng**
:tada: 

Hey all, this pull request proposes a way to incorporate modern javascript front end development workflows, built with webpacker alongside (but soon to be replacing) the rails asset pipeline.  

This is thanks to the recent native support for webpack via the official [webpacker gem](https://github.com/rails/webpacker).  

[**Click here for the TLDR**](#tldr)

### Old
```
- app
  - assets 
    - config
    - images
    - javascripts
    - stylesheets
  - controllers
  - ...
  - views 
```

### New
```
- app
  - assets
    - config
    - images
    - javascripts
    - stylesheets
  - frontend
    - admin
    - application
    - packs
  - controllers
  - ...
  - views 
```

#### Notes 
[Webpack](https://webpack.github.io/) is a front end assets module bundler.  For example it takes a collection of javascript files that depend on one another, and creates a single, minified (meaning renaming long variable names to short tokens like "a", "b" and so on to shorten the overall script length) file to load.  Once the client loads the file once, it will be cached and the client will no longer need to download front end assets again, speeding (sequential) load times.  Webpack is configured through a [`webpack.config.js`](https://webpack.js.org/concepts/configuration/#components/sidebar/sidebar.js) file.  The `Webpacker` gem has this file at `config/webpack/shared.js`.  The full configuration is split amongst other files within the `config/webpack` folder.

### Why Use this?
1. No need to use Rails clunky javascript interface (no more `form_for` tags with model binds), which also means no more `jquery-ujs` (_todo_)
2. Following best practices for modern frontend develpment in terms of tooling (webpack, es6 transpilation via babel, preact, minification) 
3. Still leverage the ability to have server side rendering for purely static html components
4. Ability to create two webpack builds for `main` site and `/admin` side.

### Wait, two builds?
Since one bundle is createed for the site, no javascript or css components for the `/admin` portion of the site should be included.  It would be best to read some of the [documentation](https://github.com/rails/webpacker#paths), or:

## TLDR;
Webpack creates a "pack" (or `bundle`) for each file in `app/frontend/packs/`.  These can be referenced in a layout/view file via some [helper functions](https://github.com/rails/webpacker#inside-views) provided by the `webpacker` gem.

## Important
New dependencies are added.  Follow through the `README.md` after merging this commit into your local machine.

### CODE REVIEW

**Comment by mohammedri** [comment]
Did @ajyang99 @rysharm and @lilcser work on this as well?

**Comment by rperryng** [comment]
@mo, No, they weren't expected to.  This involved a lot of time researching modern javascript development architectures in a rails app, and was a lot of trivial time spent tinkering with configurations to play nicely with the asset build pipeline.  A lot of this work wouldn't really make sense for the others as no one else on the team has any front end web dev experience, let alone configuring build tools and getting rails and webpack to play nice

But to get an idea of what this pr has helped me with so far:
The current state of the home page has a navbar implemented in react, written in es6 syntax and compiled into a single cached bundle downloaded once for the site.  There is also another part of the same site that is pulling information from the database and rendered serversider since it has no dynamic components.  The minimized payload is only **18.4kb**.  For reference, the tuneme website is about 405 kb, to be fair most of it is in images, but some of the two biggest files on their page are two css files that total to around 40kb.

**Comment by mohammedri** [comment]
Awesome, sweet job @rperryng 
I will do a review and 🎩 tomorrow, but since the others didn't work on this, I would like @ajyang99 @lilcser to review this code as well and give a thumbs up when you guys approve =)

**Comment by rperryng** [comment]
Merging and closing.  Unfortunately we don't have anyone with enough time or knowledge on this kind of system to continue waiting.

**Comment by chrisjluc** [comment]
You should ping @dinahshi or myself to take a look next time!

**Comment by rperryng** [comment]
@chrisjluc hey! definitely will next time.  It just didn't really make sense even for this PR, most of it is generated from adding the `webpacker` gem, and the rest is just tinkering and build config stuff that wouldn't make sense unless you're comfortable with webpack and the rails asset pipeline anyway
